### PR TITLE
Reader: enable shuffling inside every row group

### PR DESCRIPTION
**Pull request by chongxiaoc**
- Allow shuffle inside every single row group in reader.
- Also use a shuffle seed to generate reproducible shuffle results across: single row shuffle, row groups shuffle in ventilator and sharding.

### CODE REVIEW

**Comment by chongxiaoc** [comment]
Attach some benchmark result using synthetic dataset with PyTorch:
- Experiment setup: `BatchedDataLoader + make_batch_reader()`, batch_size=50000, shuffle_buffer_size=1000000, thread_pool, 10 workers.
- Datasets from 100M rows to 1.6B rows are tested. 
- Compare throughput of shuffle in dataloader and shuffle in reader .

Synthetic Dataset PyTorch Throughput

Using shuffle in reader is expected to generate higher throughput since multiple workers are shuffling in parallel.

fyi @selitvin

File: petastorm/arrow_reader_worker.py [hunk 1]
```diff
@@ -100,13 +100,17 @@ def __init__(self, worker_id, publish_func, args):
...
+        self._shuffle_seed = args[10]
 
         if self._ngram:
             raise NotImplementedError('ngrams are not supported by ArrowReaderWorker')
 
         # We create datasets lazily in the first invocation of 'def process'. This speeds up startup time since
         # all Worker constructors are serialized
         self._dataset = None
+        if self._shuffle_seed:
+            np.random.seed(seed=self._shuffle_seed)
```
Comment (selitvin) [inline petastorm/arrow_reader_worker.py:113]: This uses global state, right? This might create conflict with other users of the random generator, running on this or other threads.If we want to setup seed, I suggest we create a seeded random number generator and use it.
```python
random_state = np.random.RandomState(42)
random_state.permutations(10)
```

File: petastorm/reader.py [hunk 2]
```diff
@@ -201,11 +204,12 @@ def make_reader(dataset_url,
 def make_batch_reader(dataset_url_or_urls,
                       schema_fields=None,
                       reader_pool_type='thread', workers_count=10,
+                      shuffle_rows=False, shuffle_seed=None,
                       shuffle_row_groups=True, shuffle_row_drop_partitions=1,
                       predicate=None,
                       rowgroup_selector=None,
                       num_epochs=1,
-                      cur_shard=None, shard_count=None, shard_seed=None,
```
Comment (selitvin) [inline petastorm/reader.py:208]: Can we keep both seeds for now for the next couple of versions and add a deprecation warning? That way we don't break compatibility but give users a warning.

File: petastorm/reader.py [hunk 3]
```diff
@@ -512,7 +521,7 @@ def _filter_row_groups(self, dataset, row_groups, predicate, rowgroup_selector,
         :param cur_shard: An int denoting the current shard number used. Each node should
                        pass in a unique partition number in the range [0, shard_count).
         :param shard_count An int denoting the number of reader shards
-        :param shard_seed: If not None: random seed to shuffle row groups for data sharding.
+        :param shuffle_seed: If not None: random seed to shuffle row groups for data sharding.
```
Comment (selitvin) [inline petastorm/reader.py:524]: Should we just call this `seed` since it used for multiple purposes?

File: petastorm/tests/test_parquet_reader.py [hunk 4]
```diff
@@ -220,3 +220,41 @@ def test_string_partition(reader_factory, tmpdir, partition_by):
         row_ids_batched = [row.id for row in reader]
     actual_row_ids = list(itertools.chain(*row_ids_batched))
     assert len(data) == len(actual_row_ids)
+
+
+def test_shuffle_per_row_group(scalar_dataset):
+    """Check if every row group is shuffled."""
+    for reader_factory in [make_batch_reader, make_reader]:
+        with reader_factory(scalar_dataset.url, reader_pool_type='dummy',
+                            shuffle_rows=True, shuffle_row_groups=False) as reader:
```
Comment (selitvin) [inline petastorm/tests/test_parquet_reader.py:229]: Do you think we need to test that shuffle_rows=False work? My guess it's covered somewhere else, just wanted to make sure. Say we hardcode it to =True in make_reader/make_batch_reader: would other tests start failing?

**Review by selitvin — COMMENTED** [review]
Looks good! Just few minor comments. Also, please update release notes.

File: petastorm/arrow_reader_worker.py (hunk 1 above)
Comment (chongxiaoc) [reply petastorm/arrow_reader_worker.py:113]: Let me test. I was not sure if process pool needs to recreate a new random state.

File: petastorm/tests/test_parquet_reader.py (hunk 4 above)
Comment (chongxiaoc) [reply petastorm/tests/test_parquet_reader.py:229]: I think `shuffle_rows=False` is default case, which is tested in other existing reader tests. Those tests compare output order and length.

File: petastorm/arrow_reader_worker.py (hunk 1 above)
Comment (chongxiaoc) [reply petastorm/arrow_reader_worker.py]: addressed.

File: docs/release-notes.rst [hunk 5]
```diff
@@ -7,7 +7,7 @@ Release notes
 
 Release 0.11.6 (unreleased)
 ===========================
-
+- `PR 767 <https://github.com/uber/petastorm/pull/767>`_: Reader: enable shuffling inside every row group.
```
Comment (selitvin) [inline docs/release-notes.rst:10]: Can you please add just a bit more info? What's the benefit and which arg to use to enable it. Also, please mention deprecated argument.

**Review by selitvin — APPROVED** [review]

File: docs/release-notes.rst (hunk 5 above)
Comment (chongxiaoc) [reply docs/release-notes.rst:10]: addressed
